### PR TITLE
[inductor] Fix ConstructorMoverPass creating mixed-device pointwise ops

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #176164

ConstructorMoverPass moves CPU scalar graph inputs to CUDA so they can
be consumed by cudagraph-partition triton kernels. In
find_movable_constructors, a user node was treated as a safe "stop
point" (no further propagation needed) if all its inputs were CPU
scalars or on the target GPU device—but without checking that the user
itself was on the GPU.

Fix: require is_on_target_device(user) before treating a user as a
stop point, so CPU ops keep their original CPU inputs.

Authored with Claude.

Differential Revision: [D94933286](https://our.internmc.facebook.com/intern/diff/D94933286)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

File: test/inductor/test_cudagraph_trees.py [hunk 1]
```diff
@@ -3817,6 +3817,45 @@ def forward(self, x):
...
+            for _ in range(5):
+                output = compiled_model(x)
+                loss = criterion(output, torch.randint(0, 4, (4,)).cuda())
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+
+            eager_out = model(x)
+            compiled_out = compiled_model(x)
+            self.assertEqual(compiled_out, eager_out)
```
Comment (BoyuanFeng) [inline test/inductor/test_cudagraph_trees.py:3866]: should we check the number of graph partitions here? iiuc there should be 3 partitions (2 for gpu and 1 for cpu)

File: torch/_inductor/fx_passes/post_grad.py [hunk 2]
```diff
@@ -1796,6 +1796,41 @@ def add_cpu_inp(node):
...
+        Returns (per-scalar result nodes, set of all created nodes).
+        """
+        with graph.inserting_before(insert_after._next):
+            unsqueezed = [
+                graph.call_function(torch.ops.aten.unsqueeze.default, (s, 0))
+                for s in scalars
+            ]
+            concatenated = graph.call_function(
+                torch.ops.aten.cat.default, (unsqueezed,)
+            )
```
Comment (BoyuanFeng) [inline torch/_inductor/fx_passes/post_grad.py]: could we use torch.stack? Then we don't need unsqueeze + cat

File: torch/_inductor/fx_passes/post_grad.py [hunk 3]
```diff
@@ -1796,6 +1796,41 @@ def add_cpu_inp(node):
...
+            copied = graph.call_function(
```
Comment (BoyuanFeng) [inline torch/_inductor/fx_passes/post_grad.py]: Please use non-blocking copy to avoid cuda sync. Example:

```
x = torch.tensor(1, device="cpu", pin_memory=True)
y = x.to("cuda", non_blocking=True)
```

File: torch/_inductor/fx_passes/post_grad.py (hunk 2 above)
Comment (desertfire) [reply torch/_inductor/fx_passes/post_grad.py]: This is refactoring of the existing code, but sure, I can do that.

File: torch/_inductor/fx_passes/post_grad.py (hunk 3 above)
Comment (desertfire) [reply torch/_inductor/fx_passes/post_grad.py]: I think that `True` in `(concatenated, target_device, True)` means non_blocking. Again, this is a refactoring.

File: test/inductor/test_cudagraph_trees.py (hunk 1 above)
Comment (desertfire) [reply test/inductor/test_cudagraph_trees.py:3866]: Yes, I can add that. It's actually 2 for cpu and 1 for gpu.

File: torch/_inductor/fx_passes/post_grad.py (hunk 2 above)
Comment (desertfire) [reply torch/_inductor/fx_passes/post_grad.py]: Nop, actually fx is decomposed at this point. We can't use aten.stack here.

**Comment by BoyuanFeng** [comment]
> routing CPU tensors through CUDA partitions

Graph partition checks the node device here and exclude non-gpu nodes. If cpu tensors are passed into cuda partition, does it mean the logic here is wrong? Or maybe we should check both node's inputs and outputs?

https://github.com/pytorch/pytorch/blob/764350935dcc573221dd3fa8e29fd16abc641a62/torch/_inductor/scheduler.py#L6590-L6591

**Review by BoyuanFeng — APPROVED** [review]
